src/ucp/core/ucp_am.c

```diff
@@ -441,8 +485,14 @@ ucp_am_bcopy_pack_data(void *buffer, ucp_request_t *req, size_t length)
         /* Pack user header to the end of message/fragment */
         user_hdr = UCS_PTR_BYTE_OFFSET(buffer, payload_length);
         ucp_am_pack_user_header(user_hdr, req);
+        ucp_am_put_user_header(req);
     }
 
+    /* After packing the header this flag 
+     * need to be cleared so we don't copy it again 
+     * (when sending in multi fragments for example). */
+    req->send.msg_proto.am.flags &= ~UCP_AM_SEND_FLAG_COPY_HEADER;
```

Comment:
maybe do not reset flags and use other pack callback for middle fragments?

Reply:
There is actually different pack functions for first fragment and middle fragments
ucp_am_bcopy_pack_args_first
ucp_am_bcopy_pack_args_mid

For this reason maybe this line is redundant , but still I prefer to keep it to follow the invariant: if copy is not needed set this flag off.
Maybe update the description instead what do you think?

Reply:
see https://github.com/openucx/ucx/pull/8173/files#r885055092
Invariant here is that the request was created by API call with `UCP_AM_SEND_FLAG_COPY_HEADER`, my point is a "copy" function should be called only once instead of checking the flag every time. For example, you can add debug flag `UCP_AM_SEND_FLAG_HEADER_COPIED` to assert it in other places, but resetting the flag looks redundant operation to me.

Reply:
"but resetting the flag looks redundant operation to me"
In this case maybe, but not always.
There are flows that it's required, for example when request can go back to pending queue.

Reply:
In this case the header pointer could be self-sufficient, https://github.com/openucx/ucx/pull/8173#discussion_r885047384
If you see that the buffer was allocated from the pool, then this is a copy of header.

Reply:
I didn't understand the comment...
"If you see that the buffer was allocated from the pool" you have a way to determine if some buffer was allocated from a specific mpool? teach me this magic, it's can change things...

Reply:
it semms like we can't use lowest bit in this case since user pointer can be unaligned, but probably we can use highest bits. If not then IMO it should be 2 different flags

no magic, lowest bit could work if we control aligment for this pointer (see req_id or ep_id for example)
higest bits are typically free due to current implementation of virtual memory space [x86_64 example](https://01.org/linuxgraphics/gfx-docs/drm/x86/x86_64/mm.html)